### PR TITLE
ddl, model: support for dist-reorg on partitioned tables

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/41144

Problem Summary:
Where normal tables were supported previously, partitioned tables are supported in this PR (in addition, for lighting-merge each partition wants to be processed on the same node).

Simple implementation process
```
--------------------- controlWriteTableRecord
                                   | 
                     getRunningPhysicalTableMetas
                                   |
                                                phyTblMetaCh                 
                     sendPhysicalTableMetas   -------------->  concurrencyCnt splitPhysicalTableToBackfillJobs
                                   | 
                     checkReorgJobFinished（original code）
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -1814,46 +1817,48 @@ func (w *worker) addTableIndex(t table.Table, reorgInfo *reorgInfo) error {
...
+		return nil, nil, errors.Trace(err)
+	}
+	return getTableRange(reorg.d.jobContext(reorg.Job.ID), reorg.d, t.GetPartition(pid), currentVer.Ver, reorg.Job.Priority)
+}
+
+func getNextPartitionInfo(reorg *reorgInfo, t table.PartitionedTable, currPhysicalTableID int64) (int64, kv.Key, kv.Key, error) {
 	pi := t.Meta().GetPartitionInfo()
 	if pi == nil {
-		return true, nil
+		return 0, nil, nil, nil
```
Comment (Benjamin2037) [inline ddl/index.go:1850]: if 0 is a special value? could we define it to a const?

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/job_table_test.go [hunk 2]
```diff
@@ -604,7 +715,7 @@ func TestGetTasks(t *testing.T) {
 		se1 := ddl.NewSession(tk1.Session())
 		require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/NotifyBeginTxnCh", `return(2)`))
 		var bJobs1 []*ddl.BackfillJob
-		bJobs1, err1 = ddl.GetAndMarkBackfillJobsForOneEle(se1, 1, jobID1, uuid, instanceLease)
+		bJobs1, err1 = ddl.GetAndMarkBackfillJobsForOneEle(se1, 1, jobID1, uuid, 1, instanceLease)
```
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/job_table_test.go:718]: PhysicalTableID

File: ddl/index.go (hunk 1 above)
Comment (zimulala) [reply ddl/index.go:1836]: The fact that tableID has a value of 0 is itself illegal

File: ddl/constant.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ const (
 		ddl_job_id bigint not null,
 		ele_id bigint not null,
 		ele_key blob,
-		store_id bigint,
+		ddl_physical_tid bigint,
```
Comment (Defined2014) [inline ddl/constant.go:58]: Is there any risk here? Changed the meta of the table. And maybe we need change it back later.

File: ddl/backfilling.go [hunk 4]
```diff
@@ -61,12 +61,15 @@ const (
...
-	retrySQLTimes       = 10
+	InstanceLease                = 1 * time.Minute
+	updateInstanceLease          = 25 * time.Second
+	genTaskBatch                 = 4096
+	genPhysicalTableTaskBatch    = 256
+	minGenTaskBatch              = 1024
+	minGenPhysicalTableTaskBatch = 64
+	minDistTaskCnt               = 64
+	retrySQLTimes                = 10
+	retrySQLInterval             = 300 * time.Millisecond
```
Comment (Defined2014) [inline ddl/backfilling.go:72]: What's the difference between L76?

File: ddl/constant.go (hunk 3 above)
Comment (zimulala) [reply ddl/constant.go:58]: No, it won't be used later, but will be put into meta if needed

File: ddl/backfilling.go [hunk 5]
```diff
@@ -688,6 +700,7 @@ func getBatchTasks(t table.Table, reorgInfo *reorgInfo, kvRanges []kv.KeyRange,
 			logutil.BgLogger().Info("[ddl] get backfill range task, get reverse key failed", zap.Error(err))
 		} else {
 			logutil.BgLogger().Info("[ddl] get backfill range task, change end key",
+				zap.Int64("pID", pID), zap.Int64("pTbl", phyTbl.GetPhysicalID()), zap.Bool("mergeTIdx", reorgInfo.mergingTmpIdx),
```
Comment (tangenta) [inline ddl/backfilling.go:703]:
```suggestion
				zap.Int64("pID", pID), zap.Int64("pTbl", phyTbl.GetPhysicalID()), zap.Bool("mergeTmpIdx", reorgInfo.mergingTmpIdx),
```

File: ddl/dist_owner.go [hunk 6]
```diff
@@ -16,84 +16,311 @@ package ddl
...
+	if len(pTblMetas) == 0 {
+		bfJM := &BackfillJobRangeMeta{PhyTblID: currPID, StartKey: reorgInfo.StartKey, EndKey: reorgInfo.EndKey}
+		phyTblMetas = append(phyTblMetas, bfJM)
+		physicalTIDs = append(physicalTIDs, bfJM.PhyTblID)
+	} else {
+		for _, pMeta := range pTblMetas {
+			phyTblMetas = append(phyTblMetas, pMeta)
+			if pMeta.PhyTblID > currPID {
+				currPID = pMeta.PhyTblID
+			}
```
Comment (Defined2014) [inline ddl/dist_owner.go:102]: maybe could use `mathutil.Max` instead of it.

File: ddl/dist_owner.go [hunk 7]
```diff
@@ -16,84 +16,311 @@ package ddl
...
+			if pMeta.ID > currBfJobID {
+				currBfJobID = pMeta.ID
```
Comment (Defined2014) [inline ddl/dist_owner.go:104]: ditto

File: ddl/dist_owner.go [hunk 8]
```diff
@@ -16,84 +16,311 @@ package ddl
...
+				return
+			}
+			bfJM := &BackfillJobRangeMeta{PhyTblID: pID, PhyTbl: pTbl, StartKey: startKey, EndKey: endKey}
+			sJobCtx.phyTblMetaCh <- bfJM
+			currPhysicalID = pID
+
+			physicalTIDs = append(physicalTIDs, pID)
+		}
+	} //else {
+	//	//nolint:forcetypeassert
```
Comment (Defined2014) [inline ddl/dist_owner.go:185]: Why remain comments here?

File: ddl/dist_owner.go [hunk 9]
```diff
@@ -16,84 +16,311 @@ package ddl
...
+func (dc *ddlCtx) sendPhysicalTableMetas(reorgInfo *reorgInfo, t table.Table, sJobCtx *splitJobContext, runningPTblMetas []*BackfillJobRangeMeta) {
+	var err error
+	physicalTIDs := make([]int64, 0, distPhysicalTableConcurrency)
+	defer func() {
+		if err != nil {
+			logutil.BgLogger().Info("[ddl] send physical table ranges to split failed", zap.Int64("jobID", reorgInfo.Job.ID),
+				zap.Stringer("ele", reorgInfo.currElement), zap.Int64s("phyTblIDs", physicalTIDs), zap.Error(err))
+			sJobCtx.cancel()
+		} else {
+			logutil.BgLogger().Info("[ddl] send physical table ranges to split finished", zap.Int64("jobID", reorgInfo.Job.ID),
```
Comment (Defined2014) [inline ddl/dist_owner.go:126]: It's for debug or we will release this code? Maybe we could combine two logs.

File: ddl/job_table.go [hunk 10]
```diff
@@ -740,9 +741,28 @@ func GetAndMarkBackfillJobsForOneEle(s *session, batch int, jobID int64, uuid st
...
 
-		bJobs, err = GetBackfillJobs(se, BackfillTable,
-			fmt.Sprintf("(exec_ID = '' or exec_lease < '%v') and ddl_job_id = %d order by ddl_job_id, ele_key, ele_id limit %d",
-				leaseStr, jobID, batch), "get_mark_backfill_job")
+		getJobsSQL := fmt.Sprintf("(exec_ID = '' or exec_lease < '%v') and ddl_job_id = %d order by ddl_job_id, ele_id, ele_key limit %d",
+			leaseStr, jobID, batch)
+		if pTblID != getJobWithoutPartition {
+			if pTblID == 0 {
+				rows, err := s.execute(context.Background(),
+					fmt.Sprintf("select ddl_physical_tid from mysql.%s group by ddl_job_id, ele_id, ele_key, ddl_physical_tid having sum(length(exec_id)) = 0 or max(exec_lease) < '%s' order by ddl_job_id, ele_key, ele_id, ddl_physical_tid limit 1",
```
Comment (tangenta) [inline ddl/job_table.go:749]:
```suggestion
					fmt.Sprintf("select ddl_physical_tid from mysql.%s group by ddl_job_id, ele_id, ele_key, ddl_physical_tid having max(length(exec_id)) = 0 or max(exec_lease) < '%s' order by ddl_job_id, ele_key, ele_id, ddl_physical_tid limit 1",
```

**Comment by zimulala** [comment]
PTAL @tangenta

**Review by tangenta — APPROVED** [review]

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**